### PR TITLE
[M] Reverted product/content link from activation key and environments (CANDLEPIN-437)

### PR DESCRIPTION
**Pull request by Ceiu**
- ActivationKey no longer references products directly, instead referencing products by Red Hat product ID
- Environment no longer references content directly, instead referencing content by Red Hat content ID
- Removed the ActivationKey and Environment entity remapping logic from org refresh and product/content update flows
- Internal entity model API improvements, cleanup, and changes
- Removed some long-unused code

### CODE REVIEW

**Comment by Ceiu** [comment]
This is the first part of the changes to perform mapping resolution in the case where a mapping error is detected during refresh and import. This PR focuses on removing the churn around environments and activation keys and reverting the product and content references back to RHIDs from strict UUID references. Due to the critical nature of this change, a lot of attention should be paid to the operations surrounding environment content and activation key product usage.

Additionally, this PR focuses on updating the model APIs a bit to encourage using ID references rather than objects, while still permitting pulling children objects as necessary (specifically the owner from activation key or environment). This is less critical for Candlepin behavior, but should help make object manipulation cleaner and easier to maintain.

**Comment by wottop** [comment]
Looks like an issue here:

org.candlepin.spec.OwnerProductResourceSpecTest shouldAllowDeletingAProductAssociatedWithAnActivationKey() FAILED



  org.candlepin.invoker.client.ApiException: Message: Internal Server Error

  HTTP response code: 500

  HTTP response body: {

    "displayMessage" : "Runtime Error could not execute statement at org.mariadb.jdbc.export.ExceptionFactory.createException:282",

    "requestUuid" : "10f36dac-e1da-45a4-81d8-f55ea05fcca4"

  }

  HTTP response headers: {connection=[close], content-type=[application/json], date=[Tue, 15 Nov 2022 15:07:43 GMT], server=[Apache-Coyote/1.1], transfer-encoding=[chunked], vary=[Accept-Encoding], x-candlepin-request-uuid=[10f36dac-e1da-45a4-81d8-f55ea05fcca4], x-version=[4.2.10-1, 4.2.10-1]}

      at app//org.candlepin.spec.OwnerProductResourceSpecTest.shouldAllowDeletingAProductAssociatedWithAnActivationKey(OwnerProductResourceSpecTest.java:648)

**Review by JoshAlbrecht23 — CHANGES_REQUESTED** [review]
Looks good, just had some small comments.

File: src/test/java/org/candlepin/controller/PoolManagerTest.java [hunk 1]
```diff
@@ -1093,6 +1093,9 @@ public void testBatchRevokeCleansUpCorrectPoolsWithSourceEnt() {
     @SuppressWarnings({ "unchecked", "rawtypes" })
     @Test
     public void testEntitleWithADate() throws Exception {
+        // TODO: Fix this test with proper mocks -- the current mocks were hiding a bug in the
+        // test where we were sending the wrong product IDs in, and still got the correct result.
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply src/test/java/org/candlepin/controller/PoolManagerTest.java:1097]: The offending tests that I spotted were already fixed as part of this PR, so the TODO is referring to the general approach/usage of mocks.

We can add a card if one doesn't already exist, but I fear it'll be doomed to stay in the backlog for all eternity.

File: src/main/java/org/candlepin/model/OwnerProductCurator.java [hunk 2]
```diff
@@ -1052,60 +1048,127 @@ public void removeOwnerProductReferences(Owner owner, Collection<String> product
...
+                    .forEach(elem -> productIds.add((String) elem));
+            }
 
-                log.info("{} activation key product(s) removed", count);
+            // Delete the entries
+            // Impl note: at the time of writing, JPA doesn't support doing this operation without
+            // interacting with the objects directly. So, we're doing it with native SQL to avoid
+            // even more work here.
+            String sql = "DELETE FROM cp_activation_key_products akp " +
+                "WHERE akp.key_id = :key_id AND product_id = :product_id";
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply src/main/java/org/candlepin/model/OwnerProductCurator.java:1113]: I'll have to run through the query planner again to double check, but IIRC, due to how Hibernate caches all the individual queries until it does a flush, it ended up being faster to do them individually like that.

[the remark this answers is not in the document]
Comment (Ceiu) [reply src/main/java/org/candlepin/model/OwnerProductCurator.java:1113]: nm, I was mistaken. Fixed.

**Review by JoshAlbrecht23 — APPROVED** [review]

**Comment by Ceiu** [comment]
Rebased; and removed the very short-lived environment-content lookups and checks from the OrphanCleanupJob and ContentCurator.

**Comment by wottop** [comment]
lgtm.

**Pull request merged**